deltacat/compute/compactor/model/round_completion_info.py

```diff
@@ -31,7 +31,7 @@ def of(
         return rci
 
     @property
-    def high_watermark(self) -> Union[Dict[PartitionLocator, int], int]:
```

Comment:
can we make PartitionLocator hashable instead?

Reply:
Agreed - all locators are intended to be hashable, but looks like we just didn't override `__hash__` previously. 

This is one of the use cases for the `Locator` `digest()` method. All of our models that inherit from locator should actually override `__eq__` and `__hash__` to use the `digest` of the locator.

Reply:
On second though, `digest()` wouldn't be ideal because it is (1) expensive due to an underlying reliance on generating a SHA1 hash and (2) only delivers a point-in-time hash of the mutable locator. I think we should capture a longer term improvement to our models to either convert them all to immutable locators (i.e. inherit from `frozendict` instead of `dict`), or at least make them convertable to immutable locators. 

For the short-term, I think we can capture the same intent here of telegraphing the type of key being passed in, and recording additional information related to this key, by making a new `HighWatermark` class that keeps a `Dict[str, Tuple[int, PartitionLocator]]` under the hood where the key is the canonical string and the value is a pair of `(high_watermark_stream_position, partition_locator_for_canonical_string)`.

One way to do this is by giving `HighWatermark` a method like `append(stream_position: int, partition_locator: PartitionLocator)`.

Reply:
Thanks for the ideas. Just updated this pr; high watermark is wrapped into a class, using its canonical_string as key.

Reply:
This however, will still call the `digest()`. I think it's fine for now, the max number of calls throughout the compaction is around `O(deltas)`; In case of incremental compaction, it should be just a few calls.